### PR TITLE
2d Mapping with Matrix Gaps

### PR DESCRIPTION
**Pull request by ctjet**
This enables WLED in 2D mode to support gaps and offsets of panels in the matrix. Panels may now be spaced and ordered however the user sees fit.

Screenshot 2022-11-19 175952

### Changes

wled00/cfg.cpp
- No longer store matrix dimensions with panel units, instead, store the total count of panels
- Store an X and Y offset in pixels from top left of matrix for each panel. (uint8_t)

wled00/FX_2Dfcn.cpp
- No longer calculate matrix serpentine, bottom, right, and vertical. This functionality is now in 2D settings webpage. Instead use already set panel offsets and just calculate pixel offsets per panel relative to that.

wled00/data/settings_2D.htm
- Add "matrix generator" field. Use this field to populate the panels array and new "Number of panels" field with the correct values for a new matrix.
- Add offsets for each panel from the top left of the matrix. These can be manually tweaked to allow for gaps and weird spacings.

Other supporting changes were made for above functionality.

### Notes
I am by no means a Javascript or Web developer. All of the JS and HTML, as well as my layout of the 2D settings page, should be checked and potentially tweaked.

This should not be merged until it is decided how to handle custom maps with gaps. This functionality currently can balloon the size of the Custom Map array as every unused pixel takes up 2 bytes. This is a problem especially for ESP8266.

I am not sure if this breaks the config when upgrading to this version due to the config changes. Must be tested.

### CODE REVIEW

**Comment by blazoncek** [comment]
Looks slick.

**Comment by blazoncek** [comment]
Each iteration looks better. 👏 
Adding unequal panel support may be your next step and then it will be almost perfect.

**Comment by ctjet** [comment]
Added per panel w/h, still looking into a fix for a few bugs

File: wled00/FX.h [hunk 1]
```diff
@@ -809,22 +806,24 @@ class WS2812FX {  // 96 bytes
...
-      panelW,
       matrixWidth,
       matrixHeight;
 
     typedef struct panel_bitfield_t {
-      bool bottomStart : 1; // starts at bottom?
-      bool rightStart  : 1; // starts on right?
-      bool vertical    : 1; // is vertical?
-      bool serpentine  : 1; // is serpentine?
+      unsigned int xOffset  : 8; // x offset relative to the top left of matrix in LEDs
```
Comment (blazoncek) [inline wled00/FX.h:816]: You may want to move these four variables out of bitfield to allow lower CPU overhead.

**Comment by blazoncek** [comment]
@ctjet can you, please, pull recent commits and resolve conflicts?

**Comment by ctjet** [comment]
@blazoncek should be updated, do I need to do anything special for 0.14?

**Comment by blazoncek** [comment]
> do I need to do anything special for 0.14?

No. It just needs to be in line with recent main.

File: wled00/FX.h [hunk 2]
```diff
@@ -654,10 +654,9 @@ class WS2812FX {  // 96 bytes
       timebase(0),
       isMatrix(false),
 #ifndef WLED_DISABLE_2D
-      hPanels(1),
-      vPanels(1),
-      panelH(8),
-      panelW(8),
+      panels(1),
+      matrixWidth(DEFAULT_LED_COUNT),
```
Comment (blazoncek) [inline wled00/FX.h:658]: This set-up is not 2D, please change it that both dimensions are >1.

File: wled00/FX.h [hunk 3]
```diff
@@ -808,20 +807,24 @@ class WS2812FX {  // 96 bytes
...
     #define WLED_MAX_PANELS 64
     uint8_t
-      hPanels,
-      vPanels;
+      panels;
 
     uint16_t
-      panelH,
-      panelW;
+      matrixWidth,
```
Comment (blazoncek) [inline wled00/FX.h:813]: These two are no longer part of `strip`.

File: wled00/FX.h (hunk 1 above)
Comment (blazoncek) [reply wled00/FX.h:817]: I am still waiting for these 4 to get removed from bitfield.

File: wled00/FX_2Dfcn.cpp [hunk 4]
```diff
@@ -43,8 +43,17 @@ void WS2812FX::setUpMatrix() {
 
   // isMatrix is set in cfg.cpp or set.cpp
   if (isMatrix) {
-    Segment::maxWidth  = hPanels * panelW;
-    Segment::maxHeight = vPanels * panelH;
+    // calculate width dynamically because it will have gaps
+    matrixWidth = 0;
```
Comment (blazoncek) [inline wled00/FX_2Dfcn.cpp:47]: You will need to revert to `Segment::maxWidth` and `Segment::maxHeight`.

Do similar for the code below.

**Review by blazoncek — CHANGES_REQUESTED** [review]
There were some changes upstream that require you to update your code. It may not work properly otherwise.

I am also wondering why did you include `html_simple.h` as its source didn't change.

**Comment by blazoncek** [comment]
Maybe it is not ready after all.
@ctjet how do you plan to arrange panels? Always by entering x and y offsets?
That is awkward IMO.

**Comment by blazoncek** [comment]
Got it working now.
I'm still not sure entering offsets is a good UX.
It might be better to add graphical UI at some point, but that is beyond my skills.

**Review by Aircoookie — APPROVED** [review]

**Review by blazoncek — APPROVED** [review]

File: wled00/FX.h (hunk 1 above)
Comment (blazoncek) [reply wled00/FX.h]: Did it myself.

**Comment by blazoncek** [comment]
@softhack007 are you ok to merge this?

**Review by softhack007 — APPROVED** [review]
@blazoncek just did a quick code review, however 2D is not really my field ...
I think its good to merge 👍

**Pull request merged**